### ISSUE TITLE
Suggestion: Ability to list twitter trends 

### ISSUE THREAD

**Issue by hillz2** [ShellShoccar-jpn/kotoriotoko#4]
Your script is very useful for me but it's lacking one thing, that is to be able to list twitter trends. It would be great if you could add that function to your script. Thanks

**Comment by richmikan** [comment]
Here you are.
You want this command ["BIN/twtrends.sh"](https://github.com/ShellShoccar-jpn/kotoriotoko/blob/master/BIN/twtrend.sh), don't you?

**Comment by hillz2** [comment]
Thank you for getting back to me, sorry for the late reply, I tested the `twtrends.sh` and it worked really well on my manjaro linux, however when I try it on openwrt there are two errors:
```
root@AngelBeats:/mnt/usb2/kotoriotoko/BIN# ./twtrend.sh 23424846
./twtrend.sh: line 25: getconf: not found
twtrend.sh: OpenSSL command is not found.
```
I understand the openssl part is used to generate random strings, I can use this on openwrt instead:
```
< /dev/urandom tr -dc _A-Z-a-z-0-9 | head -c${1:-5};echo;
```
Can you give me the basic command to get the twitter trends via cURL / wget so that I can build my own script that will work with openwrt ?

**Comment by richmikan** [comment]
Well, I don't think it's easy to run kotoriotoko on OpenWrt because OpenWrt has been designed for embedded use and it omits some commands due to resource limitation and maybe not satisfy POSIX specification. But I want you to try to run kotoriotoko on it.

At first, maybe you can remove the line 25, so I recommend it. What the line is for is to reset the PATH variable. But it's not necessary to reset it unless your system change its value.

And unfortunatally, "openssl" is absolutely required. It is used for generating not only random string but also HMAC-SHA1 for OAuth 1.0a. So you have to install OpenSSL or LibreSSL into OpenWrt. It is possible to do it, isn't it?

**Comment by hillz2** [comment]
Okay I can install openssl on openwrt with
``` 
opkg install openssl-untils
```
The problem now is that openwrt doesn't support `getconf`. I have commented line 25 and it results in the following error:
[https://s9.postimg.org/ya5pm5yi7/screenshot-2017-07-17_19-30-56.png](url)

**Comment by hillz2** [comment]
I finally managed to get `twtrends.sh` to work with openwrt but I have to edit the file, here's the complete script: https://pastebin.com/5bbfWgPg
These are the lines I edited:

- Line 25 I commented out that export command
- Line 186 and 341 I changed the `randmstr` so that openwrt can execute the script perfectly

I also commented out the line:
```
export PATH="$(command -p getconf PATH)${PATH+:}${PATH-}"
``` 
on these files:
- UTL/urlencode
- TOOL/calclock
- TOOL/self

**Comment by richmikan** [comment]
Thank you for showing your console. Now, I'll tell you how to fix kotoriotoko program codes.

## 1) Remove all lines which contain the following

```
export PATH="$(command -p getconf PATH)${PATH+:}${PATH-}"
```

It is unnecessary for almost of all hosts because these codes are to satisfy POSIX strictly. It is rather harmful for OSs which are not compliant with POSIX.

## 2) Replace all lines which contain the following line in BIN/*.sh

```
randmstr=$("$CMD_OSSL" rand 8 | od -A n -t x4 -v | sed 's/[^0-9a-fA-F]//g')
```

↓

```
randmstr=$("$CMD_OSSL" rand 8 -hex)
```

OpenWrt probably doesn’t even have od command, this command must be had by All of POSIX compliant OSs. So I suggest the way which doesn't depend on the command.

Good luck!

**Comment by hillz2** [comment]
Yeah I got it working now, it would be great if all of your scripts support openwrt as well, I'm willing to help you edit them if you'd like, just let me know if you want to further support openwrt as well.

**Comment by richmikan** [comment]
I'm interested that how much OpenWrt compliants with POSIX. So, plese copy and paste and run the following commands. And report the result.

```
which which >/dev/null 2>&1 || {
  which() {
    command -v "$1" 2>/dev/null | grep '^/' || {
      echo 'which: not found' 1>&2 && (exit 1)
    }
  }
}

for cmd in admin alias ar asa at awk basename batch bc bg c99 cal cat cd cflow chgrp chmod chown cksum cmp comm command compress cp crontab csplit ctags cut cxref date dd delta df diff dirname du echo ed env ex expand expr false fc fg file find fold fort77 fuser gencat get getconf getopts grep hash head iconv id ipcrm ipcs jobs join kill lex link ln locale localedef logger logname lp ls m4 mailx make man mesg mkdir mkfifo more mv newgrp nice nl nm nohup od paste patch pathchk pax pr printf prs ps pwd qalter qdel qhold qmove qmsg qrerun qrls qselect qsig qstat qsub read renice rm rmdel rmdir sact sccs sed sh sleep sort split strings strip stty tabs tail talk tee test time touch tput tr true tsort tty type ulimit umask unalias uname uncompress unexpand unget uniq unlink uucp uudecode uuencode uustat uux val vi wait wc what who write xargs yacc zcat; do
  res=$(which $cmd 2>/dev/null)
  case $res in '') res='(N/A)';; esac
  printf '%-10s %s\n' $cmd $res
done
```

The words which are enumarated after "for cmd in" are the all commands which POSIX compatible OS must have.

**Comment by hillz2** [comment]
Here's the result: https://pastebin.com/PMFdC8uz

**Comment by richmikan** [comment]
Thanks! Hmm, some commands are omitted, just as I suspected...

I'd like to take measures soon.

**Comment by hillz2** [comment]
Yeah, there's not a lot in openwrt, let me know if you want to support this platform as well

**Comment by richmikan** [comment]
I have a question to you.
If you rewrite the beginning of "twtrend.sh" as follows, will it work properly on OpenWrt?

```
# === Initialize shell environment ===================================
set -u
umask 0022
export LC_ALL=C
type getconf >/dev/null 2>&1 &&
export PATH="$(command -p getconf PATH)${PATH+:}${PATH-}"
export UNIX_STD=2003  # to make HP-UX conform to POSIX
```

The point is the line "`type getconf >/dev/null 2>&1 &&`" above the PATH definition line.

**Comment by hillz2** [comment]
Yes, no problem with that but please note that `curl` in openwrt doesn't support the option `--compress` so you have to get rid of that option

**Comment by richmikan** [comment]
Really? please let me know the curl version.

```
$ curl --version
```

**Comment by hillz2** [comment]
```
root@AngelBeats:/mnt/usb2/kotoriotoko/BIN# curl --version                                       
curl 7.40.0 (arm-openwrt-linux-gnu) libcurl/7.40.0 PolarSSL/1.3.14                              
Protocols: file ftp ftps http https             
Features: IPv6 Largefile SSL
```
Some packages in openwrt are compiled minimally so even though some are at the latest version they might not support some options

**Comment by richmikan** [comment]
Thanks. I'll be careful about that.

**Comment by richmikan** [comment]
Finally, let me know your OpenWrt's version.

**Comment by hillz2** [comment]
OpenWrt Chaos Calmer, there should be no problem between openwrt version barrier breaker up to the latest bleeding edge "Designated Driver"

**Comment by hillz2** [comment]
On LEDE it should have no problem as well since LEDE and OpenWrt both are practically the same

**Comment by richmikan** [comment]
Thanks for your information!

**Comment by richmikan** [comment]
I improved all of the kotoriotoko's code formally to support OpenWrt. I want you to try it.

1. Back up your CONFIG/COMMON.SHLIB.
2. Install the latest file set of kotoriotoko.
3. Copy CONFIG/COMMON.SHLIB you backed up into the latest one's CONFIG directory.

**Comment by hillz2** [comment]
I've tested all the scripts in the `BIN` directory, here are some of the errors I found:

1. You forgot to remove the `od` command in `/UTL/base64` so when I execute `./getbtwid` it shows:
```
/tmp/kotoriotoko/UTL/base64: line 486: od: not found
```
2. When I execute `./stwsrch.sh` it shows the following errors:
```
root@AngelBeats:/tmp/kotoriotoko/BIN# ./stwsrch.sh 
/tmp/kotoriotoko/UTL/mktemp: line 289: bc: not found
/tmp/kotoriotoko/UTL/mktemp: line 289: od: not found
/tmp/kotoriotoko/UTL/mktemp: line 289: bc: not found
Usage   : stwsrch.sh [options] <keyword> [keyword ...]
Options :
          -u <user_ID>[,user_ID...]|--follow=<user_ID>[,user_ID...]
          -l <lat>,<long>[,<...>]  |--locations=<lat>,<long>[,<...>]
          -v                       |--verbose
          --rawout=<filepath_for_writing_JSON_data>
          --rawonly
          --timeout=<waiting_seconds_to_connect>
Version : 2017-07-18 02:39:39 JST
```
3. When I execute `./tweet.sh` without any contents it's just stuck, when I do `./tweet.sh "test message"` it executes perfectly. It's probably better if the script shows an error when no message is specified
4. `./twmediup.sh` shows the following errors:
```
root@AngelBeats:/tmp/kotoriotoko/BIN# ./twmediup.sh linux.png 
/tmp/kotoriotoko/UTL/mktemp: line 289: bc: not found
/tmp/kotoriotoko/UTL/mktemp: line 289: od: not found
/tmp/kotoriotoko/UTL/mktemp: line 289: bc: not found
/tmp/kotoriotoko/UTL/mime-make: line 1158: bc: not found
/tmp/kotoriotoko/UTL/mime-make: line 1158: od: not found
/tmp/kotoriotoko/UTL/mime-make: line 1158: bc: not found
id=887080779381055489
ex=86400
root@AngelBeats:/tmp/kotoriotoko/BIN#
```
5. English isn't my native language and I don't understand what `twplsrch.sh` does, but I guess this one works fine:
```
root@AngelBeats:/tmp/kotoriotoko/BIN# ./twplsrch.sh -g jakarta indonesia
twplsrch.sh: API error(4): Passed in granularity must be city or neighborhood.
```
6. `twvideoup.sh` returns these errors:
```
root@AngelBeats:/tmp/kotoriotoko/BIN# ./twvideoup.sh /tmp/Linux\ Pronunciation-5IfHm6R5le0.mp4 
/tmp/kotoriotoko/UTL/mktemp: line 289: bc: not found
/tmp/kotoriotoko/UTL/mktemp: line 289: od: not found
/tmp/kotoriotoko/UTL/mktemp: line 289: bc: not found
/tmp/kotoriotoko/UTL/mime-make: line 1158: bc: not found
/tmp/kotoriotoko/UTL/mime-make: line 1158: od: not found
/tmp/kotoriotoko/UTL/mime-make: line 1158: bc: not found
id=887083794783584256
ex=86400
root@AngelBeats:/tmp/kotoriotoko/BIN#
```
That's all the errors.

**Comment by richmikan** [comment]
Hmm, that's difficult to resolve the problems.. X-(
I'll do my best.

Let me know that OpenWrt has mktemp command and base64 command. (Report the result of the following)

```
$ type mktemp
$ type base64
```

**Comment by hillz2** [comment]
```
mktemp is /bin/mktemp
base64: not found
```
But I can install `base64` with the following command:
```
opkg install coreutils-base64
```
root@AngelBeats:/tmp/kotoriotoko/BIN# type base64
base64 is a tracked alias for /usr/bin/base64
```

**Comment by richmikan** [comment]
Which directory has the base64 command installed into?

```
$ type base64
```

And also let me know the value of your PATH variable.

```
$ echo $PATH
```

**Comment by hillz2** [comment]
There are some packages that you need to install to use kotoriotoko on openwrt since these packages don't come pre-installed:
```
opkg install curl coreutils-base64
```
Here are the results:
```
root@AngelBeats:/tmp/kotoriotoko/BIN# type base64
base64 is a tracked alias for /usr/bin/base64
root@AngelBeats:/tmp/kotoriotoko/BIN# echo $PATH
/usr/bin:/usr/sbin:/bin:/sbin
```

**Comment by richmikan** [comment]
Thanks.

Can you install bc command, too? bc command is also one of POSIX commands, and it is required by mime-make command.

**Comment by hillz2** [comment]
Yes I can:
```
root@AngelBeats:/tmp/kotoriotoko/BIN# opkg install bc
Installing bc (1.06.95-1) to root...            
Downloading http://downloads.openwrt.org/chaos_calmer/15.05.1/sunxi/generic/packages/packages/bc_1.06.95-1_sunxi.ipk.
Installing libreadline (6.3-1) to root...       
Downloading http://downloads.openwrt.org/chaos_calmer/15.05.1/sunxi/generic/packages/base/libreadline_6.3-1_sunxi.ipk.
Configuring libreadline.                        
Configuring bc.         
root@AngelBeats:/tmp/kotoriotoko/BIN# type bc
bc is /usr/bin/bc
```
Holy cow this distro is so limited

**Comment by richmikan** [comment]
Perhaps, you can also install od command, can't you?

**Comment by hillz2** [comment]
Yes, to wrap up on openwrt you have to install the following packages:
```
opkg install curl coreutils-od bc coreutils-base64
```
The `curl` one might be skipped since there's `wget` but I haven't tested your scripts with `wget` yet

**Comment by richmikan** [comment]
OK, I'm relieved.

I think OpenWrt gives us the omitted POSIX commands by its package management system when we required them.

**Comment by hillz2** [comment]
I tested `"BIN/btw*.sh"` they all works fine now but:
1. `stwsrch.sh` gives me this:
```
root@AngelBeats:/tmp/kotoriotoko/BIN# ./stwsrch.sh indonesia
The 1st response has arrived...
stwsrch.sh: Failed to access API
```
2. `twmediup.sh` and `twvideoup.sh` don't return any errors but they also don't upload the image & video I specified:
```
root@AngelBeats:/tmp/kotoriotoko/BIN# ./twmediup.sh /tmp/wall.jpg 
id=887098336989167616
ex=86400
root@AngelBeats:/tmp/kotoriotoko/BIN# ./twvideoup.sh /tmp/Linux\ Pronunciation-5IfHm6R5le0.mp4 
id=887098695191216129
ex=86400
root@AngelBeats:/tmp/kotoriotoko/BIN#
```

**Comment by richmikan** [comment]
The latter report probably seems to work correctly. Try the following

```
$ ./tweet.sh -m 887098336989167616 Some message
$ ./tweet.sh -m 887098695191216129 Some message
```

These commands are just file uploader. So they tweet nothing. These commands are called by tweet.sh, dmtweet.sh when `-f` option is used.

**Comment by hillz2** [comment]
Okay I can do:
```
./tweet.sh -f /tmp/wall.jpg uploading wallpaper
```
to upload a file, how about `stwsrch.sh` ? it still returns the error above

**Comment by richmikan** [comment]
To resolve the former problem, let me know the result by run the following command.

```
sh -x ./stwsrch.sh indonesia
:
(Press [Ctrl]+[C] after seeing the message "The 1st response has arrived...")
```

**Comment by hillz2** [comment]
https://pastebin.com/dzHN7tzA

**Comment by richmikan** [comment]
Thank you for your reporting, but I can't find the bug point yet.
Please let me report again with the following test code.

line 495

```
} 3>&2 2>/dev/null &           # Ignore job exiting message
```

rewrite it into the following

```
} 3>&2 2>/dev/stderr &         # Ignore job exiting message
```

And run `sh -x ./stwsrch.sh indonesia` again. (Restore the debugging code after running)

I'm sorry for not warning that the log to report contains your secret codes. And Thanks for hiding them before reporting.

**Comment by hillz2** [comment]
https://pastebin.com/n8Dq6NWw

**Comment by richmikan** [comment]
? No running log between 309-495 are contained your report because of probably my mistake. I'm sorry but rewrite stwsrch.sh again.

At line 495,

```
} 3>&2 2>/dev/null &           # Ignore job exiting message
```

Rewrite it into the following

```
} 3>&2 2>/tmp/debug.log &      # Ignore job exiting message
```

Then run `sh -x ./stwsrch.sh indonesia` again and report the log (also debug.log). (Restore the debugging code after running)

**Comment by hillz2** [comment]
Here you go, this is from the `debug.log`: https://pastebin.com/g48mzrYe

**Comment by hillz2** [comment]
At line 345 I change `"$CMD_CURL" ${no_cert_curl:-} -s` to `"$CMD_CURL" ${no_cert_curl:-} -v` and it returns something like this:
```
curl -v -H Authorization: OAuth oauth_consumer_key=xxxxxxxxx,oauth_nonce=5ae731aef35d8749,oauth_signature=xxxxxxxxx,oauth_signature_method=HMAC-SHA1,oauth_timestamp=1500382768,oauth_token=xxxxxxxxx,oauth_version=1.0,track=indonesia -d track=indonesia https://stream.twitter.com/1.1/statuses/filter.json
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0> POST /1.1/statuses/filter.json HTTP/1.1
> User-Agent: curl/7.40.0
> Host: stream.twitter.com
> Accept: */*
> Authorization: OAuth oauth_consumer_key=xxxxxxxxx,oauth_nonce=5ae731aef35d8749,oauth_signature=xxxxxxxxx,oauth_signature_method=HMAC-SHA1,oauth_timestamp=1500382768,oauth_token=xxxxxxxxx,oauth_version=1.0,track=indonesia
> Content-Length: 15
> Content-Type: application/x-www-form-urlencoded
> 
} [15 bytes data]
< HTTP/1.1 401 Authorization Required
< cache-control: must-revalidate,no-cache,no-store
< connection: close
< content-length: 266
< content-type: text/html
< date: Tue, 18 Jul 2017 05:59:34 GMT
< server: tsa
< www-authenticate: OAuth realm="Firehose"
< x-connection-hash: 03ebf6158a8a75defaf2458f6836df5f
< 
{ [266 bytes data]
100   281  100   266  100    15    183     10  0:00:01  0:00:01 --:--:--   185
+ echo The 1st response has arrived...
+ printf %s\n <html>\n<head>\n<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>\n<title>Error 401 Unauthorized</title>
+ read -r+ a_hdr
```
There's **Error 401 Unauthorized**

**Comment by richmikan** [comment]
Aaaaah, look at the line 133 of debug.log. tee command has failed. One of the cause may be this.

**Comment by hillz2** [comment]
Yes the directory is empty
```
root@AngelBeats:/tmp/_stwsrch.sh.26735.E7kna22A7Vm# ls -al
drwx------    2 root     root            40 Jul 18 12:50 .
drwxrwxrwt   25 root     root           760 Jul 18 13:12 ..
```

**Comment by richmikan** [comment]
But I woder why tee command said "No such file or directory". I have never seen it because tee command doesn't read any files but just write down into some file.

**Comment by richmikan** [comment]
Oh I see. The reason is there is no directory for that file!

**Comment by hillz2** [comment]
So how should I fix it ?

**Comment by richmikan** [comment]
Let me know in which directory you run stwsrch.sh.
And let me know the result when you run `mktemp -d -t dummy` in the directory.

**Comment by hillz2** [comment]
```
root@AngelBeats:/tmp/kotoriotoko/BIN# pwd
/tmp/kotoriotoko/BIN
root@AngelBeats:/tmp/kotoriotoko/BIN# mktemp -d -t dummy
mktemp: Invalid argument
```

**Comment by richmikan** [comment]
Sorry. Probably my mistake. Then, how about the command `mktemp -d -t dummy.XXXXXX`?

**Comment by hillz2** [comment]
```
root@AngelBeats:/tmp/kotoriotoko/BIN# mktemp -d -t dummy.XXXXXX
/tmp/dummy.zoGsqx
```
It created a directory

**Comment by richmikan** [comment]
Thanks. I will continue to investigate the problem.

**Comment by richmikan** [comment]
But the problem seems to be a little complex. So I'll try to install OpenWrt into my computer. Give me a few days to investigate. Sorry.

**Comment by richmikan** [comment]
Originally, "stwsrch.sh"'s process is the most complicated of the all kotoriotoko commands.

**Comment by hillz2** [comment]
I'll wait for further info, on my manjaro it's also stuck at `"The 1st response has arrived..."` and it returns the search result only after I press `Ctrl + C`

**Comment by richmikan** [comment]
Thanks a lot. The behavior on your manjaro is correct. I want to make also OpenWrt do such a correct behavior. See you.

**Comment by richmikan** [comment]
I have installed OpenWrt and investigated. Then I got the cause of the problem.

The cause is ps command. OpenWrt's ps command is an omitted version by BusyBox, and it's incompletely compatible with POSIX. Try to type `ps -Ao ppid,pid,comm`, then OpenWrt will complain. OpenWrt's ps commands supports only "w" option while POSIX compatible ones must be support "-A" and "-o" option at least.

"stwsrch.sh" requires "-o" option supporting one because "stwsrch.sh" have to get the process ID of curl/wget which "stwsrch.sh" called to control it.

Do you have any good ideas to install the ps command that supports these options.

**Comment by hillz2** [comment]
Can't you use grep like this:
```
root@AngelBeats:/mnt/usb2/kotoriotoko/BIN# ps | grep -E "sshpass.*1096" | grep -v grep
22303 root       668 S    sshpass -p zzzzzzzzzz ssh -p 443 -vND 192.168.1.1:1096 
```

**Comment by richmikan** [comment]
Unfortunately, that idea can't resolve the problem.

"stwsrch.sh" wants ps command to let it notice PPIDs. But OpenWrt's one doesn't give the information because it doesn't support "-o PPID".

It is easy to just get all of the curl/wget process IDs. But stwsrch.sh have to get only the one it created by itself to avoid killing the curl/wget process which was created by another stwsrch.sh.

**Comment by hillz2** [comment]
I mean you can do something like this:
```
ps | grep -E "curl .*stream\.twitter\.com" | grep -v grep
```
This will grep the only process that is used by `stwsrch.sh` since it only greps `"curl .*stream\.twitter\.com"` and other curl processes won't be shown. I can't find another way to install `ps` on openwrt that will support all those commands above.

**Comment by richmikan** [comment]
Umm... That's slightly wrong. If you run two or more "stwsrch.sh"s on the same host at the same time, or if you and a friend of yours do that, stwsrch.sh will probably kill a wrong curl/wget process by mistake.

**Comment by hillz2** [comment]
Too bad I can't find any packages that can do something like that in openwrt

**Comment by richmikan** [comment]
Umm... I’m in trouble. I will think through the problem to resolve it.

**Comment by hillz2** [comment]
I'll let you know when I can come up with something, so far I don't think it's possible to install `ps` with all those options

**Comment by richmikan** [comment]
Thanks a lot! And me, too.

**Comment by hillz2** [comment]
You can do `opkg install procps-ps` and then reboot your openwrt, you'll be able to do `ps -Ao ppid,pid,comm` but the file `/tmp/_stwsrch.sh.12527.Lvi3Jgx7ANA/apires` is still not created when I run the script. There must be another problem.

**Comment by richmikan** [comment]
I could find a bug in "UTL/mktemp", which is a mktemp wrapper, thanks to you. Please replace it to [the newer version](https://github.com/ShellShoccar-jpn/kotoriotoko/blob/master/UTL/mktemp). And try stwsrch.sh again.

However, it's OpenWrt's "command"'s fault originally. According to POSIX document, the "command" command must obey the original PATH even if the current PATH variable has changed anything.

```
* At first, change PATH value
$ export PATH=(PATH_TO_KOTORIOTOKO)/UTL:$PATH

* correct behavior
$ command -v mktemp
(/usr/bin, /bin, and so on)/mktemp

* in OpenWrt case (wrong behavior)
$ command -v mktemp
(PATH_TO_KOTORIOTOKO)/UTL/mktemp
```

**Comment by hillz2** [comment]
There's a coding error in your `mktemp` file https://pastebin.com/N1BZkvLv

**Comment by richmikan** [comment]
???
my mktemp and stwsrch.sh on my OpenWrt work good. There is no such an error message.

Is your mktemp correct? Please confirm the following.

```
$ md5sum /mnt/usb2/kotoriotoko/UTL/mktemp
92d6bb07cc14d702ef42e266074f1039  mktemp

If your mktemp's code was different from the above, Reget it again with the following command.
$ wget https://raw.githubusercontent.com/ShellShoccar-jpn/kotoriotoko/master/UTL/mktemp
$ chmod +x mktemp
$ mv mktemp /mnt/usb2/kotoriotoko/UTL
```

**Comment by hillz2** [comment]
May I know your curl version on openwrt? it doesn't produce any code error anymore but on openwrt I still get 
```
root@AngelBeats:/mnt/usb2/kotoriotoko/BIN# ./stwsrch.sh indonesia
The 1st response has arrived...
stwsrch.sh: API error: Error 401 Unauthorized
```
But on my manjaro PC, it works just fine, but anyway I can use `./btwsrch.sh indonesia` so it doesn't really matter. Good work

**Comment by richmikan** [comment]
OK, I investigated.

```
$ curl --version
curl 7.40.0 (x86_64-openwrt-linux-gnu) libcurl/7.40.0 PolarSSL/1.3.14
Protocols: file ftp ftps http https
Features: IPv6 Largefile SSL
```

It seems that all version number are same as yours except CPU architecture.

**Comment by hillz2** [comment]
```
root@AngelBeats:/mnt/usb2/kotoriotoko/BIN# curl --version
curl 7.40.0 (arm-openwrt-linux-gnu) libcurl/7.40.0 PolarSSL/1.3.14
Protocols: file ftp ftps http https 
Features: IPv6 Largefile SSL
```
It's ok, I'll just close this issue since it's not even related to twitter trends anymore. Thanks for your scripts, good work.

**Comment by hillz2** [comment]
One more thing, you can probably update the `README.md` with the packages required to run the scripts on openwrt.

**Comment by richmikan** [comment]
You are right. Thanks. Please open new issue, I have a question to you to resolve it.
